### PR TITLE
Linker call graph with DCE

### PR DESCRIPTION
**Pull request by nicolasstucki**
Add the linker call graph and dead code elimination.

Missing language features
- [x] varargs: add call on `wrapArray`s and `Predef` when calling a method with varargs.
- [x] Add call `fillStackTrace` when instantiating a new exception.
- [x] Java allocated types.
- [x] Code potentially called from java.
- [x] Add calls to `Predef` module init.
- [x] Add calls to default arguments.
- [x] Add calls to initial values in traits.
- [x] Add calls to mixin constructors.
- [x] Add `@export` to mark entry points to the callgraph.
- [ ] Handle FunctionN

Infrastructure:
- [x] Create visual representation of the callgraph that can scale on number of calls.
- [x] Create compiler flag to enable DCE (currently `-link-dce`)
- [x] Create compiler flag to enable visual callgraph output (currently `-link-vis`)
- [x] Modularize code into distinct classes.
- [x] Make partest configuration to test the callgraph.

Bugs to fix:
- [x] Call to `fillStackTrace` in `ControlThrowable`
- [x] Load module of entry points
- [x] Fix call to mixin constructors.
- [x] Transitively reachable from java (link-code-from-java-3).
- [x] Fix issue with type parameters in JavaAllocatedTypes
- [ ] Fix bug with lazy vals
- [ ] Fix bug with closures

...

### CODE REVIEW

File: src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -1267,15 +1267,15 @@ class TypeComparer(initctx: Context) extends DotClass with ConstraintHandling {
...
     else
       PolyType(
-        paramNames = tpnme.syntheticLambdaParamNames(tparams1.length),
+        paramNames = tpnme.syntheticTypeParamNames(tparams1.length),
         variances = (tparams1, tparams2).zipped.map((tparam1, tparam2) =>
           (tparam1.paramVariance + tparam2.paramVariance) / 2))(
         paramBoundsExp = tl => (tparams1, tparams2).zipped.map((tparam1, tparam2) =>
           tl.lifted(tparams1, tparam1.paramBoundsAsSeenFrom(tp1)).bounds &
-          tl.lifted(tparams2, tparam2.paramBoundsAsSeenFrom(tp2)).bounds),
+            tl.lifted(tparams2, tparam2.paramBoundsAsSeenFrom(tp2)).bounds),
```
Comment (nicolasstucki) [inline src/dotty/tools/dotc/core/TypeComparer.scala]: Oh, did not notice that reformatting. I will remove all that.

Comment (nicolasstucki) [inline src/dotty/tools/dotc/core/TypeComparer.scala]: Note to self: the formatting changes where done in 3ac15f0ec0fcdbe0aacd3952ec113a71a9910af2

File: compiler/src/dotty/tools/dotc/transform/linker/types/JavaAllocatedType.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package dotty.tools.dotc.transform.linker.types
+
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Hashable
+import dotty.tools.dotc.core.Types.{PolyType, SingletonType, Type}
+
+class JavaAllocatedType(private[JavaAllocatedType] val u: Type) extends SingletonType {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/linker/types/JavaAllocatedType.scala]: I think that's just `private`

File: compiler/src/dotty/tools/dotc/transform/linker/types/JavaAllocatedType.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  /** customized hash code of this type.
+    * NotCached for uncached types. Cached types
+    * compute hash and use it as the type's hashCode.
+    */
+  def hash: Int = {
+    val underlying = u.hash
+    if (underlying == Hashable.NotCached) Hashable.NotCached
+    else if (underlying == Hashable.NotCached - 1) underlying
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/linker/types/JavaAllocatedType.scala]: There is another special value: `HashUnknown`.

File: compiler/src/dotty/tools/dotc/transform/linker/types/JavaAllocatedType.scala (hunk 2 above)
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/transform/linker/types/JavaAllocatedType.scala]: Changed

**Comment by felixmulder** [comment]
*disabled

**Comment by nicolasstucki** [comment]
@DarkDimius I'm getting the following failure while running `test`. Do you have some useful insight?
```scala
WARNING: run tests can only be run by partest, JUnit just verifies compilation: ../tests/link/stdlib-dce/../../../scala-scala/src/library/scala/runtime/RichBoolean.scala
exception while transforming def <init>(self: Boolean): Unit of class class dotty.tools.dotc.ast.Trees$DefDef # 7243857
exception while transforming (self: Boolean) extends AnyVal() with scala.runtime.OrderedProxy[Boolean] { 
  val self: Boolean
  protected def ord: scala.math.Ordering[Boolean] = scala.math.Ordering.Boolean
} of class class dotty.tools.dotc.ast.Trees$Template # 7243927
exception while transforming @scala.annotation.internal.SourceFile(
  "../tests/link/stdlib-dce/../../../scala-scala/src/library/scala/runtime/RichBoolean.scala"
) final class RichBoolean(self: Boolean) extends AnyVal() with scala.runtime.OrderedProxy[Boolean] { 
  val self: Boolean
  protected def ord: scala.math.Ordering[Boolean] = scala.math.Ordering.Boolean
} of class class dotty.tools.dotc.ast.Trees$TypeDef # 7243928
exception while transforming package runtime {
  @scala.annotation.internal.SourceFile(
    "../tests/link/stdlib-dce/../../../scala-scala/src/library/scala/runtime/RichBoolean.scala"
  ) final class RichBoolean(self: Boolean) extends AnyVal() with scala.runtime.OrderedProxy[Boolean] { 
    val self: Boolean
    protected def ord: scala.math.Ordering[Boolean] = scala.math.Ordering.Boolean
  }
  final lazy module val RichBoolean: scala.runtime.RichBoolean$ = new scala.runtime.RichBoolean$()
  @scala.annotation.internal.SourceFile(
    "../tests/link/stdlib-dce/../../../scala-scala/src/library/scala/runtime/RichBoolean.scala"
  ) final module class RichBoolean$() extends Object() { this: scala.runtime.RichBoolean.type =>}
} of class class dotty.tools.dotc.ast.Trees$PackageDef # 7243934
exception while transforming package scala {
  package runtime {
    @scala.annotation.internal.SourceFile(
      "../tests/link/stdlib-dce/../../../scala-scala/src/library/scala/runtime/RichBoolean.scala"
    ) final class RichBoolean(self: Boolean) extends AnyVal() with scala.runtime.OrderedProxy[Boolean] { 
      val self: Boolean
      protected def ord: scala.math.Ordering[Boolean] = scala.math.Ordering.Boolean
    }
    final lazy module val RichBoolean: scala.runtime.RichBoolean$ = new scala.runtime.RichBoolean$()
    @scala.annotation.internal.SourceFile(
      "../tests/link/stdlib-dce/../../../scala-scala/src/library/scala/runtime/RichBoolean.scala"
    ) final module class RichBoolean$() extends Object() { this: scala.runtime.RichBoolean.type =>}
  }
} of class class dotty.tools.dotc.ast.Trees$PackageDef # 7243935
[error] Test dotc.tests.link_dce_all failed: java.lang.AssertionError: assertion failed: data race? overwriting symbol of type ([scala$math$Ordered$$A] => ()Unit)(scala.runtime.RichBoolean#<init>),
[error] long form = TermRef(TypeRef(ThisType(TypeRef(NoPrefix,runtime)),RichBoolean),<init>)/withSig(Signature(List(),scala.runtime.BoxedUnit)) of class class dotty.tools.dotc.core.Types$TermRefWithSignature,
[error] last sym id = 4285, new sym id = 4303,
[error] last owner = trait Ordered, new owner = trait Proxy,
[error] period = callGraph at run 2, took 24.933 sec
[error]     at scala.Predef$.assert(Predef.scala:165)
[error]     at dotty.tools.dotc.core.Types$NamedType.checkSymAssign(Types.scala:1490)
[error]     at dotty.tools.dotc.core.Types$NamedType.setDenot(Types.scala:1525)
[error]     at dotty.tools.dotc.core.Types$NamedType.withDenot(Types.scala:1518)
[error]     at dotty.tools.dotc.core.Types$NamedType.withDenot(Types.scala:1516)
[error]     at dotty.tools.dotc.core.Types$TermRef$.apply(Types.scala:1899)
[error]     at dotty.tools.dotc.core.Types$NamedType$.apply(Types.scala:1859)
[error]     at dotty.tools.dotc.transform.linker.CollectSummaries$Collect$$anonfun$prepareForDefDef$1.apply(CollectSummaries.scala:180)
[error]     at dotty.tools.dotc.transform.linker.CollectSummaries$Collect$$anonfun$prepareForDefDef$1.apply(CollectSummaries.scala:178)
[error]     at scala.collection.immutable.List.foreach(List.scala:381)
[error]     at dotty.tools.dotc.transform.linker.CollectSummaries$Collect.prepareForDefDef(CollectSummaries.scala:178)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$28.apply(TreeTransform.scala:532)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$28.apply(TreeTransform.scala:532)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.mutateTransformers(TreeTransform.scala:490)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformNamed(TreeTransform.scala:965)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1178)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1171)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:136)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1170)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformSub(TreeTransform.scala:1216)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformUnnamed(TreeTransform.scala:1146)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1179)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1171)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:136)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1170)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformNamed(TreeTransform.scala:979)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1178)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1171)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:136)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1170)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.dotty$tools$dotc$transform$TreeTransforms$TreeTransformer$$transformStat$1(TreeTransform.scala:1204)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$37.apply(TreeTransform.scala:1208)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$37.apply(TreeTransform.scala:1208)
[error]     at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:52)
[error]     at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:68)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformStats(TreeTransform.scala:1208)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformUnnamed(TreeTransform.scala:1158)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1179)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1171)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:136)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1170)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.dotty$tools$dotc$transform$TreeTransforms$TreeTransformer$$transformStat$1(TreeTransform.scala:1206)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$37.apply(TreeTransform.scala:1208)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$37.apply(TreeTransform.scala:1208)
[error]     at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:52)
[error]     at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:68)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformStats(TreeTransform.scala:1208)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transformUnnamed(TreeTransform.scala:1158)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1179)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer$$anonfun$transform$2.apply(TreeTransform.scala:1171)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:136)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.transform(TreeTransform.scala:1170)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.macroTransform(TreeTransform.scala:551)
[error]     at dotty.tools.dotc.transform.TreeTransforms$TreeTransformer.run(TreeTransform.scala:469)
[error]     at dotty.tools.dotc.transform.TreeTransforms$MiniPhase$class.run(TreeTransform.scala:164)
[error]     at dotty.tools.dotc.transform.linker.CollectSummaries.run(CollectSummaries.scala:151)
[error]     at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:279)
[error]     at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:277)
[error]     at scala.collection.immutable.List.map(List.scala:273)
[error]     at dotty.tools.dotc.core.Phases$Phase$class.runOn(Phases.scala:277)
[error]     at dotty.tools.dotc.transform.linker.CollectSummaries.runOn(CollectSummaries.scala:22)
[error]     at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:75)
[error]     at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:72)
[error]     at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
[error]     at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
[error]     at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply$mcV$sp(Run.scala:72)
[error]     at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:66)
[error]     at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:66)
[error]     at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:76)
[error]     at dotty.tools.dotc.Run.compileUnits(Run.scala:66)
[error]     at dotty.tools.dotc.Run.compileSources(Run.scala:63)
[error]     at dotty.tools.dotc.Run.compile(Run.scala:47)
[error]     at dotty.tools.dotc.Driver.doCompile(Driver.scala:26)
[error]     at dotty.tools.dotc.Driver.process(Driver.scala:124)
[error]     at dotty.tools.dotc.Driver.process(Driver.scala:93)
[error]     at dotty.tools.dotc.CompilerTest.compileArgs(CompilerTest.scala:281)
[error]     at dotty.tools.dotc.CompilerTest.compileFile(CompilerTest.scala:107)
[error]     at dotty.tools.dotc.CompilerTest$$anonfun$compileFiles$1.apply(CompilerTest.scala:210)
[error]     at dotty.tools.dotc.CompilerTest$$anonfun$compileFiles$1.apply(CompilerTest.scala:208)
[error]     at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
[error]     at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
[error]     at dotty.tools.dotc.CompilerTest.compileFiles(CompilerTest.scala:208)
[error]     at dotty.tools.dotc.CompilerTest.runFiles(CompilerTest.scala:220)
[error]     at dotc.tests$$anonfun$link_dce_all$1.apply(tests.scala:196)
[error]     at dotc.tests$$anonfun$link_dce_all$1.apply(tests.scala:194)
[error]     at scala.collection.immutable.List.foreach(List.scala:381)
[error]     at dotc.tests.link_dce_all(tests.scala:194)
[error]     ...
```

**Comment by DarkDimius** [comment]
@nicolasstucki, strange stacktrace. Looks like it's updating an existing denotation and this shouldn't happen.

**Comment by smarter** [comment]
> I'm getting the following failure while running test. Do you have some useful insight?

From the code, this is what you do in a loop with `sym` some fixed constructor and `decl` being every mixin primary constructor of `sym.owner`:
```scala
val initRef = ref(NamedType(sym.owner.typeRef, decl.name, decl.denot))
```
`NamedType` ends up calling `TermRef#apply` which will try to create or get from cache a `TermRef` from:
- The prefix, which here is always fixed
- The name of the constructor, which is always `<init>`
- The signature of the denotation of the constructor, which is very likely going to be `Signature(List(),scala.runtime.BoxedUnit)`

So you're very likely to just fetch the same `TermRef` multiple time, except that `TermRef#apply` will also call `withDenot` on the fetched `TermRef` and set the denot you passed, this fails because you pass denots with different owners (since they all come from different mixins).

I have no clue what this is trying to achieve (please add comments!), so I can't tell you how to fix it.

**Comment by nicolasstucki** [comment]
Added the outer map substitution.

**Pull request closed without merging**